website/docs/concepts/wasm-bindgen.md

```diff
@@ -0,0 +1,239 @@
+Document Object Model (DOM) works on the web. When types are imported using `wasm-bindgen` you can
+also add attributes that describe it's inheritance.
+
+In Rust this inheritance is simulated using [`Deref`](https://doc.rust-lang.org/std/ops/trait.Deref.html)
+and [`AsRef`](https://doc.rust-lang.org/std/convert/trait.AsRef.html) traits. An example of this
+might help; so say you have three types `A`, `B`, and `C` where `C` extends `B` which extends `A`.
+When importing these types the `#[wasm-bindgen]` macro will implement the `Deref` and `AsRef`
+traits in the following way:
+
+- `C` can `Deref` to `B`
+- `B` can `Deref` to `A`
+- `C` can be `AsRef` to `B`
+- Both `C` & `B` can be `AsRef` to `A`
+
+These implementations allow you to call a method from `A` on an instance of `C` and to use `C` as if
+it was `&B` or `&A`.
+
+It's important to note that every single type imported using `#[wasm-bindgen]` has the same root type,
+you can think of it as the `A` in the example above, this type is [`JsValue`](#jsvalue) which has
+it's own section
+below.
+
+_[extends section in The `wasm-bindgen` Guide](https://rustwasm.github.io/docs/wasm-bindgen/reference/attributes/on-js-imports/extends.html)_
+
+### [`JsValue`](https://rustwasm.github.io/wasm-bindgen/api/wasm_bindgen/struct.JsValue.html)
+
+This is a representation of an object owned by JavaScript, this is a root catch-all type for `wasm-bindgen`.
+Any other type that comes from `wasm-bindgen` is a `JsValue` and this is because JavaScript doesn't have
+a strong type system so any function that accepts a variable `x` doesn't define it's type so `x` can be
+an valid JavaScript value; hence `JsValue`. So when you are working with imported functions or types that
+accept a `JsValue` then any value is _technically_ valid.
+
+`JsValue` can be accepted by a function but that function may still only expect certain types and this
+can lead to panics - so when using raw `wasm-bindgen` APIs check the documentation of the JavaScript
+being imported whether an exception will be caused if that value is not a certain type.
+
+_[`JsValue` documentation](https://rustwasm.github.io/wasm-bindgen/api/wasm_bindgen/struct.JsValue.html)._
+
+### [`JsCast`](https://rustwasm.github.io/wasm-bindgen/api/wasm_bindgen/trait.JsCast.html)
+
+Rust has a strong type system and JavaScript...doesn't 😞 So in order for Rust to maintain these
+strong types but still be convenient the web assembly group came up with a pretty neat trait `JsCast`.
+It's job is to help you move from one JavaScript "type" to another, which sounds vague but it means
+that if you have one type which you know is really another then you can use the functions of `JsCast`
+to jump from one type to the other. It's a nice trait to get to know when working with `web-sys`,
+`wasm_bindgen`, `js-sys` - you'll notice lots of types will implement `JsCast` from those crates.
```

Comment:
Perhaps we could somehow reference the `dyn Any` mechanism from the standard library here?
Not sure if we can expect many users to know about it though, so it might not really be that helpful.

Reply:
I can see why, but I don't agree. `JsCast` helps jumping between strong types, whereas `dyn Any` is more about erasure of types. I don't think of them in the same way and I would normally try and avoid `dyn Any` if possible.

Reply:
That's a good point. I was only thinking about it in terms of implementation and API methods, but semantically they're pretty different.

Reply:
I'm worried about giving people the foot gun `dyn Any` 😅